logagg_collector/collector.py

```diff
@@ -0,0 +1,926 @@
+    def _collect_log_files(self, log_files):
+        '''
+        Collect from log files in logaggfs 'logs' one by one
+        '''
+
+        L = log_files
+        # Sorted list of all the files for one pattern
+        fpaths = glob.glob(join(self.logaggfs.logs_dir, L['fpattern']))
+        fpaths = sorted(fpaths)
+
+        for f in fpaths:
+            log_files.update({'fpath': f})
+            # If last file in the list keep polling until next file arrives
+            self._collect_log_lines(log_files)
+            if not f == fpaths[-1]:
+                self.log.debug('deleting_file', f=f)
+                self._delete_file(f)
+        time.sleep(1)
+
+
+    def _collect_log_lines(self, log_file):
+        '''
+        Collects logs from logfiles, formats and puts in queue
+        '''
+        L = log_file
+        fpath = L['fpath']
+        fmtfn = L['formatter_fn']
+        formatter = L['formatter']
+
+        freader = Pygtail(fpath)
+        for line, line_info in self._iter_logs(freader, fmtfn):
+            log = self._assign_default_log_values(fpath, line, formatter)
+
+            try:
+                _log = fmtfn(line)
+                # Identify logs inside a log
+                # Like process logs inside docker logs
+                if isinstance(_log, RawLog):
+                    formatter, raw_log = _log['formatter'], _log['raw']
+                    log.update(_log)
+                    # Give them to actual formatters
+                    _log = load_formatter_fn(formatter)(raw_log)
+
+                log.update(_log)
+            except (SystemExit, KeyboardInterrupt) as e: raise
+            except:
+                log['error'] = True
+                log['error_tb'] = traceback.format_exc()
+                self.log.exception('error_during_handling_log_line', log=log['raw'])
+
+            if log['id'] == None:
+                log['id'] = uuid.uuid1().hex
+
+            log = self._remove_redundancy(log)
+            if self._validate_log_format(log) == 'failed': continue
+
+            self.queue.put(dict(log=json.dumps(log),
+                                freader=freader, line_info=line_info))
+            self.log.debug('tally:put_into_self.queue', size=self.queue.qsize())
+
+        while not freader.is_fully_acknowledged():
+            t = self.PYGTAIL_ACK_WAIT_TIME
+            self.log.debug('waiting_for_pygtail_to_fully_ack', wait_time=t)
+            time.sleep(t)
+
+
+    def _get_msgs_from_queue(self, msgs, timeout):
+        msgs_pending = []
+        read_from_q = False
+        ts = time.time()
+
+        msgs_nbytes = sum(len(m['log']) for m in msgs)
+
+        while 1:
+            try:
+                msg = self.queue.get(block=True, timeout=self.QUEUE_READ_TIMEOUT)
+                read_from_q = True
+                self.log.debug("tally:get_from_self.queue")
+
+                _msgs_nbytes = msgs_nbytes + len(msg['log'])
```

Comment:
Please do confirm that log will never get a new line character in msg['log'] if we are not sure please do use `.strip()` for msg['log']